readme.txt
asw (Analog Stick Wheel)

Asw turns the thumbstick of an Xbox controller (or any other XInput compatible controller) into a wheel-like control. This can allow for more precise control in racing games.

=========
Overview:
=========
Asw is a vJoy feeder and needs vJoy to work. vJoy is a device driver that adds a virtual game controller to the system.
Asw does 2 things.

1. Asw reads the rotational movement of the left thumbstick on the Xbox controller (or any other XInput compatible controller) and converts it to a linear value that it sends (feeds) to a vJoy axis. Or you can say that it creates an axis (on the vJoy device) that responds to the rotation of the controller analog stick.
Basically, it makes the thumbstick work kind of like a wheel.

2. Asw also reads the left and right triggers and feeds the values to vJoy.
This is useful because when the Xbox controller is used in DirectInput mode, the left and right triggers share an axis. This function allows you to have independent inputs from the triggers present on the vJoy device (each trigger will have it's own axis).

Asw was made to improve the level of precision offered by the Xbox controller in racing games.

====================
System requirements:
====================
Windows 7 or later (only tested in Win 7 and 10)
Xbox controller or any other XInput compatible controller
vJoy (tested with version v2.1.8)

=======================
How to install and use:
=======================
First, download asw from asw's github page. There are two ways you can achieve this.
Method 1) Download the entire source code package using the green download button on project’s main page (https://github.com/aswdevz/asw), all that you need to run asw will be inside the "Release" directory. 
Method 2) Download the package from the releases section (https://github.com/aswdevz/asw/releases), this will contain just the program and documentation.

Then install vJoy (vJoy is required to run asw), you can get it from here:
http://vJoystick.sourceforge.net/site/index.php/download-a-install/download
Check that vJoy has at least one virtual device using the "Configure vJoy" or "Monitor vJoy" tools (these are included with vJoy). This device should have at least these 3 controls: x axis, slider 0, slider 1. It can have more, but if one of the previously mentioned controls is missing, asw will lose the corresponding functionality. The device should be created automatically when you install vJoy. But if it's not present for some reason, create it. 
vJoy devices have numbers from 1-16. Asw will try to use the device numbered 1. You can change this behavior using the -interface command line argument.
Note: the 3 controls this feeder uses can be changed with command line arguments that are explained later in this file.

Next, run asw (asw.exe).
Asw will start working (feed inputs to vJoy) right after it's launched. You can confirm that asw is working by checking the last message in the command window, the message should be: "asw version x.x successfully started, press esc to quit".
In case asw does not start properly, try launching it from the command line so you can see the program output. The output may explain what the problem is.
You should now be able to assign asw's controls in your racing game and use them.
A guide showing how to make asw work with Assetto Corsa and Project Cars is available in the doc directory. Screenshots showing the final control configuration for these two games are also present.
Asw should work with other games as well, but only these two were tested.
The recommended way to use the controller with asw is: push the left thumbstick to the edge with your thumb, then move it along the edge.
Pressing the back button and Y button together will reset the wheel to center.
Pressing the Escape key exits asw.
You can adjust how asw works using command line arguments, explained below.

====================================
Command line arguments for asw.exe:
====================================
-interface or -i
sets the vJoy interface to use
valid vJoy interface numbers are 1-16
default is 1
example: -i 3

-controller or -c
sets the controller to use
XInput controllers are numbered 1-4 (they are actually numbered 0-3 in the code)
default is 1 (or 0 in the code)
example: -c 2

-deadzone or -d
sets the deadzone size to use for the thumbstick input in percent
valid deadzone setting is 0-100
default is 15
example: -d 5

-rotation or -r
sets the maximum number of degrees the virtual wheel can be rotated to (or wheel lock)
default is 900
example: -r 720

-bindmode or -b
Bindmode is used to assign (bind) an axis in game when it can't be done by usual means. You see, when you use this program to feed vJoy and use the controller, the system will register inputs from the vJoy virtual controller you created and the XInput controller at the same time. This means that the game will probably see inputs from both of these controllers which can mean you won't be able to assign the right one when the game prompts you for controller input while assign the control. To get past this, use this argument followed by a 1,2 or 3.
1 to make asw generate movement in the virtual wheel axis
2 to make asw generate movement in the left trigger
3 to make asw generate movement in the right trigger
That way you should be able to bind the control if it doesn't work normally.
example:
-b 2

-anykey or -a
when this is present, any key press will exit the program
if this is not present, only pressing the Escape key will exit the program
example: -a

================================
ADVANCED command line arguments:
================================
-bindmodeincrement or -bmi
Adjust the speed of movement generated by bindmode.
You probably won't need to touch this.
defaults is 128
Higher values mean faster movement
Smaller values mean slower movement
A value of 128 will make the control move from end position to end position in about 2 seconds

-bindmodemovementdirection or -bmmd
Adjusts the direction of movement generated by bindmode.
Accepts -1 and 1 only. 1 means forward direction, -1 means reverse direction.
example: -bmnd -1

-bindmodenonreversible or -bmnr
When this is used, bindmode will move the axis in one direction only. The normal behavior is reversing the direction when end of travel is reached.
example: -bmnr

-bindmoderesetwait or -bmrw
When bindmode moves the axis to the end of it's travel, bindmode will stop and wait for a bit before it starts moving the axis again. This option can set this delay. The delay is set in cycles. The default value is 125 cycles and the default length of a cycle is 8ms so the default delay is roughly 125*8ms = 1000ms = 1 second.
example: -bmrw 250
note: the above command will double the usual delay

-wheelresetbuttonsflag or -wrbf
Change the button or button combination that has to be pressed to reset the wheel to center.
This argument accepts a hexadecimal value only.
Default is: 0x8020 (back and Y buttons)
Use one of the values from the following table:
0x0001  XInput_GAMEPAD_DPAD_UP
0x0002  XInput_GAMEPAD_DPAD_DOWN
0x0004  XInput_GAMEPAD_DPAD_LEFT
0x0008  XInput_GAMEPAD_DPAD_RIGHT
0x0010  XInput_GAMEPAD_START
0x0020  XInput_GAMEPAD_BACK
0x0040  XInput_GAMEPAD_LEFT_THUMB
0x0080  XInput_GAMEPAD_RIGHT_THUMB
0x0100  XInput_GAMEPAD_LEFT_SHOULDER
0x0200  XInput_GAMEPAD_RIGHT_SHOULDER
0x1000  XInput_GAMEPAD_A
0x2000  XInput_GAMEPAD_B
0x4000  XInput_GAMEPAD_X
0x8000  XInput_GAMEPAD_Y
Or, if you want the reset to be triggered by pressing a combination of buttons, just add
the values together (using hexadecimal addition) and pass the result to this option.
example:
-wrbf C000
note: the above will make the reset happen when X and Y buttons are pressed at the same time
this is because 0x4000 + 0x8000 = 0xC000
note: the value can be passed in with the "0x" in front of the value or without it

-disablewheelreset or -dwr
The wheel will no longer reset to center if the assigned button combination is pressed.

-tickdelay or -t
Sets how long to wait (in milliseconds) between consecutive vJoy device updates.
This basically sets the delay in milliseconds between each tick/cycle (each time the main loop is run).
Setting this to 8 means that the vJoy device will get updated 1000/8 = 125 times a second.
Setting this to 25 will mean that the vJoy device will get updated 1000/25 = 40 times a second.
The default value is 8. The minimum allowed value is 1 and the maximum allowed value is 1000.
You should probably keep this at the default value.
example:
-t 17

-rightstick or -rs
Use the right thumb stick on your controller for the virtual wheel input instead of the left thumb stick.
example:
-rs

-wheelaxis or -wa
Sets the vJoy device axis that should be used as output for the virtual wheel.
Enter any value that vJoy recognizes as an axis identifier. The value should be a hexadecimal integer.
The recognized axis identifiers are at the end of this file.
Default is: 0x30
exaple 1:
-wa 0x31
exaple 2:
-wa 31
note: both examples will result in the same behavior

-lefttriggeraxis or -lta
Sets the vJoy device axis that should be used as output for the left trigger.
Enter any value that vJoy recognizes as an axis identifier. The value should be a hexadecimal integer.
The recognized axis identifiers are at the end of this file.
Default is: 0x36
exaple 1:
-wa 0x32
exaple 2:
-wa 32
note: both examples will result in the same behavior

-righttriggeraxis or -rta
Sets the vJoy device axis that should be used as output for the right trigger.
Enter any value that vJoy recognizes as an axis identifier. The value should be a hexadecimal integer.
The recognized axis identifiers are at the end of this file.
Default is: 0x37
exaple 1:
-wa 0x35
exaple 2:
-wa 35
note: both examples will result in the same behavior

-invertwheel or -iw
inverts virtual wheel axis output

-invertletfttrigger or -ilt
invert left trigger output

-invertrighttrigger or -irt
invert right trigger output

-help or -h
display help text

-version or -v
display program version

========================================
The axis identifiers in vJoy v2.1.8 are:
========================================
Axis     Value
X        0x30
Y        0x31
Z        0x32
Rx       0x33
Ry       0x34
Rz       0x35
Slider0  0x36
Slider1  0x37
Wheel    0x38
POV      0x39

=========================
Repository url + contact:
=========================
https://github.com/aswdevz/asw
aswdevz@gmail.com


changelog.txt
v1.0.3
- wheel will now reset to center by pressing the "back" + "Y" button combination (configurable)
- control assignment in project cars works better (bindmode is improved and has more options)
- from now on, asw won't start if incorrect command line arguments are used
- fixed bug that caused the invertlefttrigger option to not be recognized

v1.0.2
- made readme more informative

v1.0.1
- improved readme
- started changelog

v1.0.0
- initial release


asw/asw.h
// asw.h : this is the main header file of this program. I will include this
// into every source file. A lof of the header files included here can probably
// be moved into stdafx.h
//

#pragma once

#define VERSION_STRING "1.0.3"
//#define EXTRA_LINES

#include <stdlib.h> // used for printf and maybe other things
#include <iostream> // just used for std::system("pause");

#include <windows.h>

#define _USE_MATH_DEFINES
#include <math.h>

// the following 3 files are taken from vJoy218SDK-291116.zip
// this can be downloaded here: http://vjoystick.sourceforge.net/site/index.php/81-news/104-vjoy-2-1-5-sdk-released
#include <public.h> // needed for vjoy
#include <vjoyinterface.h> // needed for vjoy
#pragma comment(lib, "lib\\vJoyInterface.lib")

// the following 2 files are taken from: Microsoft Direct X SDK (June 2010) (DXSDK_Jun10.exe)
// this can be downloaded here: https://www.microsoft.com/en-us/download/details.aspx?id=6812
#include <Xinput.h> // needed to access the 360 controller through xinput
#pragma comment(lib, "lib\\XInput.lib")

struct s_aliaslist
{
	UINT* iInterface;
	DWORD* controller_number;
	int* thumbstick_deadzone;
	int* virtual_wheel_max_angle;
	int* bind_mode;
	bool* any_key_to_quit;
	int* bind_mode_increment;
	int* bind_mode_movement_direction;
	bool* bind_mode_non_reversible;
	int* bind_mode_reset_wait;
	int* wheel_reset_buttons_flag;
	bool* disable_wheel_reset;
	int* tick_delay;
	bool* use_right_stick;
	bool* help_detected;
	UINT* wheel_axis;
	UINT* left_trigger_axis;
	UINT* right_trigger_axis;
	LONG* invert_wheel_axis;
	LONG* invert_left_trigger;
	LONG* invert_right_trigger;
};

asw/xy_to_angle.cpp
// xy_to_angle.cpp : This file contains functions that take the X and Y values representing
// the position of the controller thumbstick and convertes them to an angle and then to
// an integer that can be sent to vjoy. Functions that help this process are also present.
//

//#define USE_ORIGINAL_DEADZONE_FUNCTION // there are two variations of the deadzone function, this switches between them

#include "stdafx.h" // for precompiled headers, must be present in every source file

#include "asw.h" // I'm including this in every source file
#include "xy_to_angle.h"

#ifdef USE_ORIGINAL_DEADZONE_FUNCTION
// the deadzone will be +1% higher with this implementation
bool in_deadzone(short x, short y, int deadzone)
{
	int x_percent = (int)(abs(x) / 327.67);
	int y_percent = (int)(abs(y) / 327.67);
	//cout << "xper= " << x_percent << " yper= " << y_percent << " total= " << (pow(x_percent, 2) + pow(y_percent, 2)) << " dz= " << pow(deadzone,2) << endl;
	if ( sqrt(pow(x_percent, 2) + pow(y_percent, 2) ) <= deadzone)
	{
		return TRUE;
	}
	return FALSE;
}

#else
// alternate deadzone function
bool in_deadzone(short x, short y, int deadzone)
{
	double x_percent = abs(x) / 327.67;
	double y_percent = abs(y) / 327.67;
	

	if ( sqrt(pow(x_percent, 2) + pow(y_percent, 2) )  <= deadzone)
	{
		//printf(">IN DEADZONE<");
		return TRUE;
	}
	return FALSE;
}
#endif

double calculate_angle_from_x_y(short x, short y)
{
	if (x == 0) // make sure we don't get a divide by zero
	{
		if (y >= 0)
			return 0.0;
		else
			return 180.0;
	}
	if (y == 0)
	{
		if (x >= 0)
			return 90.0;
		else
			return -90.0;
	}

	if (x > 0)
	{
		if (y > 0) // first quadrant
		{
			return (atan((double)x / (double)y) * 180.0 / M_PI);
		}
		if (y < 0) // fourth quadrant
		{
			return ((atan(-(double)y / (double)x) * 180.0 / M_PI) + 90.0);
		}

	}
	if (x < 0)
	{
		if (y > 0) // second quadrant
		{
			return ((-atan(-(double)x / (double)y)) * 180.0 / M_PI);
		}
		if (y < 0) // third quadrant
		{
			return ((-atan(-(double)y / -(double)x) * 180.0 / M_PI) - 90.0);
		}
	}

	return 0.0;
}

// this converts the angle of the pretend wheel to an input value in the range vjoy works with
// vjoy axis input works in this range: 0x0001 to 0x8000 (32768)
// the correct calcualtion for this conversion is: (angle + max_angle) * (vjoy_range / angle_range) + 1
// angle range is abs(min_angle) + abs(max_angle) which is equal to virtual_wheel_max_angle
// vjoy range is max_vjoy_input_value - min_vjoy_input_value = 32768 - 1 = 32767
// the result is rounded before conversion to long
long angle_to_vjoy(double angle, int virtual_wheel_max_angle)
{
	double max_angle = ((double)virtual_wheel_max_angle) / 2;
	double min_angle = -((double)virtual_wheel_max_angle) / 2;

	if (angle > max_angle)
		angle = max_angle;
	if (angle < min_angle)
		angle = min_angle;

	long finalvalue = (long)round(((angle + max_angle)*(32767.0 / ((double)virtual_wheel_max_angle)) + 1));
	//long finalvalue = (long)((angle + max_angle)*(32767.0 / ((double)virtual_wheel_max_angle)) + 1);

	return finalvalue;
}

int check_if_rotated_past_180(double current, double last, int* rotations)
{
	// the (current >= -180.0) condition can also be (current = -180.0)
	// this is because the function that prepares data for this one only
	// outputs a range of 180.0 (inclusive) to -180.0 (noninclusive)
	// the (current < 0.0) condition can also be (current <= 0.0) but this makes no difference
	// this is because when current is 0, the last condition can't ever be met
	// (current < (last - 180.0)) condition checked - it's ok
	if (((current >= -180.0) && (current < 0.0)) && (current < (last - 180.0))) // case 1
	{
		*rotations = *rotations + 1;
		//cout << current << " last " << last << endl;
		return 1;
	}

	// the (current <= 180.0) condition is correct, a value of 180.0 can appear
	// this is because the function that prepares data for this one
	// outputs a range of 180.0 (inclusive) to -180.0 (noninclusive)
	// the (current > 0.0) condition can also be (current >= 0.0) but this makes no difference // recently changed this - retest
	// this is because when current is 0, the last condition can't ever be met
	// (current > (last + 180.0)) condition checked - it's ok
	if (((current <= 180.0) && (current > 0.0)) && (current > (last + 180.0))) // case 3
	{
		*rotations = *rotations - 1;
		return -1;
	}
	return 0;
}

// there is a tiny mistake here, that does not cause issues
int clip_angle_and_rotations(int virtual_wheel_max_angle, double* angle_in, int* rotations)
{
	double right_lock = ((double)virtual_wheel_max_angle) / 2;
	double left_lock = -(((double)virtual_wheel_max_angle) / 2);

	// mistake is here: the following two lines will not assign the correct values to max_rotations and
	// min_rotations for a virtual_wheel_max_anglee of 360, 1080, 1800, 2520 etc., this is because at 360,
	// the number of times the stick can rotate past 180 (-180) is actually 0 but this code sets the maximum to 1 (-1)
	int max_rotations = (int)((right_lock + 180.0) / 360.0);
	int min_rotations = (int)((left_lock - 180.0) / 360.0);

	// the above mistake is not a problem since the following two lines will set the max_angle to
	// -180.0 if the above situation happens and a -180.0 and 1 rotation is the same as 180.0 and 0 rotations
	// as far as the code that takes the rusults of this function as input is concerned (in case the 
	// virtual_wheel_max_angle is 360 and a value of 180.0 is the input angle). A simmilar thing will happen
	// in all other cases.
	double max_angle = right_lock - (360.0 * max_rotations);
	double min_angle = left_lock + (360.0 * (-min_rotations));

	double angle = ((*angle_in) + (*rotations)*360.0);

	if (angle > ((double)virtual_wheel_max_angle) / 2)
	{
		angle = ((double)virtual_wheel_max_angle) / 2;
		*angle_in = max_angle;
		*rotations = max_rotations;
	}
	if (angle < (-((double)virtual_wheel_max_angle) / 2))
	{
		angle = -(((double)virtual_wheel_max_angle) / 2);
		*angle_in = min_angle;
		*rotations = min_rotations;
	}

	//cout << "rl = " << right_lock << " ll = " << left_lock << " maxr= " << max_rotations << " minr= " << min_rotations << " max-a " << max_angle << " min-a " << min_angle << " out= " << angle << endl;
	return 0;
}

// this calculates the real angle by taking the raw_angle which is a value from -180.0 (noninclusive)
// to 180.0 (inclusive) and adds the the number of full rotations to this value
double calculate_real_angle(double current, int virtual_wheel_max_angle, int* rotations, int* steering_lock)
{
	//	cout << "current " << current << endl;
	*steering_lock = 0;

	double angle = (current + (*rotations)*360.0);

	// this calcualtion is wrong because: clip_angle_and_rotations() function will always adjust the angle that enters this function
	// so that the angle is always at most the maximum possible value as defined by (double)virtual_wheel_max_angle) / 2
	// so these two condions will never be true, so change the conditons from (angle > ((double)virtual_wheel_max_angle) / 2)
	// to (angle = ((double)virtual_wheel_max_angle) / 2) or completely remove this because steering_lock variable
	// is not used anywhere, also this calculation can be moved to clip_angle_and_rotations()
	if (angle > ((double)virtual_wheel_max_angle) / 2)
	{
		angle = ((double)virtual_wheel_max_angle) / 2;
		*steering_lock = 1;
	}
	if (angle < (-((double)virtual_wheel_max_angle) / 2))
	{
		angle = -(((double)virtual_wheel_max_angle) / 2);
		*steering_lock = -1;
	}
	return angle;
}

LONG invert_vjoy_axis_value(LONG value, int invert_axis)
{
	if (invert_axis == 1)
		return 32769 - value;
	else
		return value;
}


asw/cmdline.cpp
// cmdline.cpp : This file contains functions that parse the command line arguments
//

#include "stdafx.h" // for precompiled headers, must be present in every source file

#include "asw.h" // I'm including this in every source file
#include "cmdline.h"

bool parse_cmdline(s_aliaslist aliaslist, int argc, char* argv[])
{
	int int_scan_val = 0;

	#ifdef EXTRA_LINES
	printf("argument count: %d\n", argc);
	#endif
	for (int i = 0; i < argc; i++)
	{
		#ifdef EXTRA_LINES
		printf("argument %d : %s\n", i, argv[i]);
		#endif
	}

	for (int i = 1; i < argc; i++)
	{
		#ifdef EXTRA_LINES
		printf(">>cmdline parse loop<<\n");
		#endif

		//std::cout << "addr " << &argv[i] << std::endl;

		if ((argcmp(argv[i], "interface") == 0) || (argcmp(argv[i], "i") == 0))
		{
			//printf("-interface detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%d", &int_scan_val) == 1)
				{
					// valid vjoy interface numbers are 1-16
					*(aliaslist.iInterface) = int_scan_val;
					i++;
				}
				else
				{
					printf("-interface value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-interface value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "controller") == 0) || (argcmp(argv[i], "c") == 0))
		{
			//printf("-controller detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%d", &int_scan_val) == 1)
				{
					// valid xinput controller number is 0-3
					// 0 corresponds to controller number 1 etc.
					*(aliaslist.controller_number) = int_scan_val-1;
					i++;
				}
				else
				{
					printf("-controller value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-controller value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "deadzone") == 0) || (argcmp(argv[i], "d") == 0))
		{
			//printf("-deadzone detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%d", &int_scan_val) == 1)
				{
					// valid deadzone setting is 0-100
					if ((int_scan_val >= 0) && (int_scan_val <= 100))
					{
						*(aliaslist.thumbstick_deadzone) = int_scan_val;
						i++;
					}
					else
					{
						printf("-deadzone value out of range, min=0, max=100\n");
						return false;
					}
				}
				else
				{
					printf("-deadzone value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-deadzone value not present\n");
				return false;
			}
			continue;
		}

		if ( (argcmp(argv[i], "rotation") == 0) || (argcmp(argv[i], "r") == 0) )
		{
			//printf("-rotation detected\n");
			if ( (i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%d", &int_scan_val) == 1)
				{
					// I should probably clip this to 0-32767 just to be sure
					*(aliaslist.virtual_wheel_max_angle) = int_scan_val;
					i++;
				}
				else
				{
					printf("-rotation value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-rotation value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "bindmode") == 0) || (argcmp(argv[i], "b") == 0))
		{
			//printf("-bindmode detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%d", &int_scan_val) == 1)
				{
					// bindmode should be 0-3
					if ((int_scan_val >= 0) && (int_scan_val <= 3))
					{
						*(aliaslist.bind_mode) = int_scan_val;
						i++;
					}
					else
					{
						printf("-bindmode value out of range, min=0, max=3\n");
						return false;
					}
				}
				else
				{
					printf("-bindmode value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-bindmode value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "anykey") == 0) || (argcmp(argv[i], "a") == 0))
		{
			//printf("-anykey detected\n");
			*(aliaslist.any_key_to_quit) = 1;
			continue;
		}

		if ((argcmp(argv[i], "bindmodeincrement") == 0) || (argcmp(argv[i], "bmi") == 0))
		{
			//printf("-bindmodeincrement detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%d", &int_scan_val) == 1)
				{
					// bindmodeincrement should be 1-8192
					if ((int_scan_val >= 1) && (int_scan_val <= 8192))
					{
						*(aliaslist.bind_mode_increment) = int_scan_val;
						i++;
					}
					else
					{
						printf("-bindmodeincrement value out of range, min=1, max=8192\n");
						return false;
					}
				}
				else
				{
					printf("-bindmodeincrement value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-bindmodeincrement value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "bindmodemovementdirection") == 0) || (argcmp(argv[i], "bmmd") == 0))
		{
			//printf("-bindmodemovementdirection detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%d", &int_scan_val) == 1)
				{
					// bindmodemovementdirection should be 1 or -1
					if ((int_scan_val == 1) || (int_scan_val == -1))
					{
						*(aliaslist.bind_mode_movement_direction) = int_scan_val;
						i++;
					}
					else
					{
						printf("-bindmodemovementdirection value out of range, allowed values: 1 and -1\n");
						return false;
					}
				}
				else
				{
					printf("-bindmodemovementdirection value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-bindmodemovementdirection value not present\n");
				return false;
			}
			continue;
		}
		
		if ((argcmp(argv[i], "bindmodenonreversible") == 0) || (argcmp(argv[i], "bmnr") == 0))
		{
			//printf("-bindmodenonreversible detected\n");
			*(aliaslist.bind_mode_non_reversible) = 1;
			continue;
		}

		if ((argcmp(argv[i], "bindmoderesetwait") == 0) || (argcmp(argv[i], "bmrw") == 0))
		{
			//printf("-bindmoderesetwait detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%d", &int_scan_val) == 1)
				{
					*(aliaslist.bind_mode_reset_wait) = int_scan_val;
					i++;
				}
				else
				{
					printf("-bindmoderesetwait value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-bindmoderesetwait value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "wheelresetbuttonsflag") == 0) || (argcmp(argv[i], "wrbf") == 0))
		{
			//printf("-wheelresetbuttonsflag detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%x", &int_scan_val) == 1)
				{
					*(aliaslist.wheel_reset_buttons_flag) = int_scan_val;
					i++;
				}
				else
				{
					printf("-wheelresetbuttonsflag value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-wheelresetbuttonsflag value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "disablewheelreset") == 0) || (argcmp(argv[i], "dwr") == 0))
		{
			//printf("-disablewheelreset detected\n");
			*(aliaslist.disable_wheel_reset) = 1;
			continue;
		}

		if ((argcmp(argv[i], "tickdelay") == 0) || (argcmp(argv[i], "t") == 0))
		{
			//printf("-tickdelay detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%d", &int_scan_val) == 1)
				{
					// tickdelay should be 1-1000
					if ((int_scan_val >= 1) && (int_scan_val <= 1000))
					{
						*(aliaslist.tick_delay) = int_scan_val;
						i++;
					}
					else
					{
						printf("-tickdelay value out of range, min=1, max=1000\n");
						return false;
					}
				}
				else
				{
					printf("-tickdelay value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-tickdelay value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "rightstick") == 0) || (argcmp(argv[i], "rs") == 0))
		{
			//printf("-rightstick detected\n");
			*(aliaslist.use_right_stick) = 1;
			continue;
		}

		if ((argcmp(argv[i], "wheelaxis") == 0) || (argcmp(argv[i], "wa") == 0))
		{
			//printf("-wheelaxis detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%x", &int_scan_val) == 1)
				{
					*(aliaslist.wheel_axis) = int_scan_val;
					i++;
				}
				else
				{
					printf("-wheelaxis value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-wheelaxis value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "lefttriggeraxis") == 0) || (argcmp(argv[i], "lta") == 0))
		{
			//printf("-lefttriggeraxis detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%x", &int_scan_val) == 1)
				{
					*(aliaslist.left_trigger_axis) = int_scan_val;
					i++;
				}
				else
				{
					printf("-lefttriggeraxis value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-lefttriggeraxis value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "righttriggeraxis") == 0) || (argcmp(argv[i], "rta") == 0))
		{
			//printf("-righttriggeraxis detected\n");
			if ((i + 1) < argc)
			{
				if (sscanf_s(argv[i + 1], "%x", &int_scan_val) == 1)
				{
					*(aliaslist.right_trigger_axis) = int_scan_val;
					i++;
				}
				else
				{
					printf("-righttriggeraxis value not present or in incorrect format\n");
					return false;
				}
			}
			else
			{
				printf("-righttriggeraxis value not present\n");
				return false;
			}
			continue;
		}

		if ((argcmp(argv[i], "invertwheel") == 0) || (argcmp(argv[i], "iw") == 0))
		{
			//printf("-invertwheel detected\n");
			*(aliaslist.invert_wheel_axis) = 1;
			continue;
		}

		if ((argcmp(argv[i], "invertlefttrigger") == 0) || (argcmp(argv[i], "ilt") == 0))
		{
			//printf("-invertlefttrigger detected\n");
			*(aliaslist.invert_left_trigger) = 1;
			continue;
		}

		if ((argcmp(argv[i], "invertrighttrigger") == 0) || (argcmp(argv[i], "irt") == 0))
		{
			//printf("-invertrighttrigger detected\n");
			*(aliaslist.invert_right_trigger) = 1;
			continue;
		}

		if ((argcmp(argv[i], "help") == 0) || (argcmp(argv[i], "h") == 0) || (argcmp(argv[i], "?") == 0))
		{
			//printf("-help detected\n");
			if (*(aliaslist.help_detected))
				printf("\n");
			*(aliaslist.help_detected) = true;
			printf("Usage: asw [-i vjoyid] [-c controller] [-d size] [-r range] [-b]\n");
			printf("           [-a] [-bmi increment] [-bmmd code] [-bmnr] [-bmrw delay]\n");
			printf("           [-wrbf buttons] [-dwr] [-t delay] [-rs] [-wa axis]\n");
			printf("           [-lta axis] [-rta axis] [-iw] [-ilt] [-irt] [-h] [-v]\n");
			printf("\n");
			printf("Options:\n");
			printf("  -i vjoyid       vJoy interface to use [1-16]\n");
			printf("  -c controller   controller to use [1-4]\n");
			printf("  -d size         deadzone for the thumbstick input in percent\n");
			printf("  -r range        number of degrees the virtual wheel can be rotated to\n");
			printf("  -b              run program in bindmode\n");
			printf("  -a              make any key press quit the program\n");
			printf("  -bmi increment  adjust the speed of bindmode\n");
			printf("  -bmmd value     changes the movement direction in bindmode\n");
			printf("  -bmnr           bindmode movement will be one way only\n");
			printf("  -bmrw delay     wait time for when bindmode reaches end of travel\n");
			printf("  -wrbf buttons   sets the wheel reset to center button or combination\n");
			printf("  -dwr            disables wheel reset to center button or combination\n");
			printf("  -t delay        delay between consecutive vJoy device updates\n");
			printf("  -rs             use right thumbstick as input for the virtual wheel\n");
			printf("  -wa axis        vJoy device axis used as output for the virtual wheel\n");
			printf("  -lta axis       vJoy device axis used as output for the left trigger\n");
			printf("  -rta axis       vJoy device axis used as output for the right trigger\n");
			printf("  -iw             invert virtual wheel axis output\n");
			printf("  -ilt            invert left trigger output\n");
			printf("  -irt            invert right trigger output\n");
			printf("  -h              display this help text\n");
			printf("  -v              display program version, current ver: %s\n", VERSION_STRING);
			continue;
		}

		if ((argcmp(argv[i], "version") == 0) || (argcmp(argv[i], "v") == 0))
		{
			//printf("-version detected\n");
			if (*(aliaslist.help_detected))
				printf("\n");
			*(aliaslist.help_detected) = true;
			printf("asw version %s\n", VERSION_STRING);
			continue;
		}

		// add extra comparisons here

		printf("unrecognized argument: %s\n", argv[i]);
		return false;
	}

	return true;
}

int argcmp(char* arg, char* cmp)
{
	int returnval = 1;
	
	int cmp1size = strlen(cmp) + 1 + 1;
	char* cmp1 = new char[cmp1size]();
	strcpy_s(cmp1, cmp1size, "-");
	strcat_s(cmp1, cmp1size, cmp);
	returnval = strcmp(arg, cmp1);
	delete[] cmp1;
	if (returnval == 0)
		return returnval;

	cmp1size = strlen(cmp) + 1 + 1;
	cmp1 = new char[cmp1size]();
	strcpy_s(cmp1, cmp1size, "/");
	strcat_s(cmp1, cmp1size, cmp);
	returnval = strcmp(arg, cmp1);
	delete[] cmp1;
	if (returnval == 0)
		return returnval;

	cmp1size = strlen(cmp) + 1 + 2;
	cmp1 = new char[cmp1size]();
	strcpy_s(cmp1, cmp1size, "--");
	strcat_s(cmp1, cmp1size, cmp);
	returnval = strcmp(arg, cmp1);
	delete[] cmp1;
	if (returnval == 0)
		return returnval;

	return returnval;
}